apps/opencs/view/render/terrainselection.cpp

```diff
@@ -0,0 +1,114 @@
+#include "terrainselection.hpp"
+
+#include <osg/Group>
+
+#include "../../model/world/cellcoordinates.hpp"
+
+#include "worldspacewidget.hpp"
+
+namespace
+{
+    const int cellSize {ESM::Land::REAL_SIZE};
+    const int landSize {ESM::Land::LAND_SIZE};
+    const int landTextureSize {ESM::Land::LAND_TEXTURE_SIZE};
+}
+
+CSVRender::TerrainSelection::TerrainSelection(const CSMWorld::CellCoordinates& coords, const ESM::Land& esmLand, osg::Group* parentNode)
+    :mCoords {coords}, mEsmLand {esmLand}, mParentNode {parentNode}
+{
+    mBaseNode = new osg::PositionAttitudeTransform{};
+    mBaseNode->setPosition(osg::Vec3d(coords.getX() * cellSize, coords.getY() * cellSize, 10.0));
+    activate();
+}
+
+CSVRender::TerrainSelection::~TerrainSelection()
+{
+    deactivate();
+}
+
+void CSVRender::TerrainSelection::select(const WorldspaceHitResult& hit)
+{
+    deselect();
+
+    if (isInCell(hit)) {
+        addToSelection(hit.worldPos);
+    }
+    update();
+}
+
+void CSVRender::TerrainSelection::toggleSelect(const WorldspaceHitResult& hit)
+{
+    if (isInCell(hit)) {
+        toggleSelection(hit.worldPos);
+        update();
+    }
+}
+
+void CSVRender::TerrainSelection::activate()
+{
+    mParentNode->addChild(mBaseNode);
+}
+
+void CSVRender::TerrainSelection::deactivate()
+{
+    mParentNode->removeChild(mBaseNode);
+}
+
+CSVRender::TerrainSelection::TerrainPos CSVRender::TerrainSelection::toTextureCoords(osg::Vec3d worldPos) const
```

Comment:
Non-primitive types should be passed by const reference.

Reply:
Vec3d is a small class, only consisting of three doubles, so I assumed it would be OK for passing by value. I will change it if this is a problem though.